remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java

```diff
@@ -0,0 +1,241 @@
+import org.apache.kafka.common.log.remote.storage.RemoteStorageManager;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.Objects;
+
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.transfer.TransferManager;
+import com.amazonaws.services.s3.transfer.TransferManagerBuilder;
+import com.amazonaws.services.s3.transfer.Upload;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * AWS S3 RemoteStorageManager.
+ *
+ * Stores files as {@code {topic}-{partition}/00000000000000000123.{log|index|timeindex}.{uuid}}.
+ */
+public class S3RemoteStorageManager implements RemoteStorageManager {
+    private static final Logger log = LoggerFactory.getLogger(S3RemoteStorageManager.class);
+
+    private AwsClientBuilder.EndpointConfiguration endpointConfiguration = null;
+
+    private String bucket;
+    private AmazonS3 s3Client;
+    private TransferManager transferManager;
+
+    public S3RemoteStorageManager() {
+    }
+
+    // for testing
+    S3RemoteStorageManager(final AwsClientBuilder.EndpointConfiguration endpointConfiguration) {
+        Objects.requireNonNull(endpointConfiguration, "endpointConfiguration must not be null");
+        this.endpointConfiguration = endpointConfiguration;
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+        Objects.requireNonNull(configs, "configs must not be null");
+
+        final S3RemoteStorageManagerConfig config = new S3RemoteStorageManagerConfig(configs);
+        this.bucket = config.s3BucketName();
+
+        AmazonS3ClientBuilder s3ClientBuilder = AmazonS3ClientBuilder.standard();
+        if (this.endpointConfiguration == null) {
+            s3ClientBuilder = s3ClientBuilder.withRegion(config.s3Region());
+        } else {
+            s3ClientBuilder = s3ClientBuilder.withEndpointConfiguration(endpointConfiguration);
+        }
+
+        // It's fine to pass null in here.
+        s3ClientBuilder.setCredentials(config.awsCredentialsProvider());
+
+        s3Client = s3ClientBuilder.build();
+        transferManager = TransferManagerBuilder.standard().withS3Client(s3Client).build();
+    }
+
+    @Override
+    public RemoteLogSegmentContext copyLogSegment(final RemoteLogSegmentId remoteLogSegmentId,
+                                                  final LogSegmentData logSegmentData) throws RemoteStorageException {
+        Objects.requireNonNull(remoteLogSegmentId, "remoteLogSegmentId must not be null");
+        Objects.requireNonNull(logSegmentData, "logSegmentData must not be null");
+
+        final long baseOffset = Log.offsetFromFileName(logSegmentData.logSegment().getName());
```

Comment:
Since this implementation will be outside of the Apache Kafka codebase eventually, will it have access to `Log`?

Reply:
Of course we can add the runtime dependency on `core`, but the dependency surface is so limited that it's easier to just copy the implementation of these couple of functions. Done.

Reply:
Thanks.